chart/chart.go

```diff
@@ -0,0 +1,214 @@
+	preDocs      string = "docs/"
+)
+
+// Chart represents a complete chart.
+//
+// A chart consists of the following parts:
+//
+// 	- Chart.yaml: In code, we refer to this as the Chartfile
+// 	- templates/*: The template directory
+// 	- README.md: Optional README file
+// 	- LICENSE: Optional license file
+// 	- hooks/: Optional hooks registry
+//	- docs/: Optional docs directory
+//
+// Packed charts are stored in gzipped tar archives (.tgz). Unpackaged charts
+// are directories where the directory name is the Chartfile.Name.
+//
+// Optionally, a chart might also locate a provenance (.prov) file that it
+// can use for cryptographic signing.
+type Chart interface {
+	// Chartfile resturns a *Chartfile for this chart.
+	Chartfile() *Chartfile
+	// Dir returns a directory where the chart can be accessed.
+	Dir() string
+
+	// Close cleans up a chart.
+	Close() error
+}
+
+type dirChart struct {
+	chartfile *Chartfile
+	dir       string
+}
+
+func (d *dirChart) Chartfile() *Chartfile {
+	return d.chartfile
+}
+
+func (d *dirChart) Dir() string {
+	return "."
+}
+
+func (d *dirChart) Close() error {
+	return nil
+}
+
+type tarChart struct {
+	chartfile *Chartfile
+	tmpDir    string
+}
+
+func (t *tarChart) Chartfile() *Chartfile {
+	return t.chartfile
+}
+
+func (t *tarChart) Dir() string {
+	return "."
+}
+
+func (t *tarChart) Close() error {
+	// Remove the temp directory.
+	return os.RemoveAll(t.tmpDir)
+}
+
+// LoadDir loads an entire chart from a directory.
+//
+// This includes the Chart.yaml (*Chartfile) and all of the manifests.
+//
+// If you are just reading the Chart.yaml file, it is substantially more
+// performant to use LoadChartfile.
+func LoadDir(chart string) (Chart, error) {
+	if fi, err := os.Stat(chart); err != nil {
+		return nil, err
+	} else if !fi.IsDir() {
+		return nil, fmt.Errorf("Chart %s is not a directory.", chart)
+	}
+
+	cf, err := LoadChartfile(filepath.Join(chart, "Chart.yaml"))
+	if err != nil {
+		return nil, err
+	}
+
+	c := &dirChart{
+		chartfile: cf,
+		dir:       chart,
+	}
+
+	return c, nil
+}
+
+// Load loads a chart from a chart archive.
+//
+// A chart archive is a gzipped tar archive that follows the Chart format
+// specification.
+func Load(archive string) (Chart, error) {
+	if fi, err := os.Stat(archive); err != nil {
+		return nil, err
+	} else if fi.IsDir() {
+		return nil, errors.New("cannot load a directory with chart.Load()")
+	}
+
+	raw, err := os.Open(archive)
+	if err != nil {
+		return nil, err
+	}
+	defer raw.Close()
+
+	unzipped, err := gzip.NewReader(raw)
+	if err != nil {
+		return nil, err
+	}
+	defer unzipped.Close()
+
+	untarred := tar.NewReader(unzipped)
+	c, err := loadTar(untarred)
+	if err != nil {
+		return c, err
+	}
+
+	cf, err := LoadChartfile(filepath.Join(c.tmpDir, ChartfileName))
+	if err != nil {
+		return c, err
+	}
+	c.chartfile = cf
+	return c, nil
+}
+
+func loadTar(r *tar.Reader) (*tarChart, error) {
+	td, err := ioutil.TempDir("", "chart-")
```

Comment:
Is there any reason why we need to use a temp directory (and the associated management) instead of using a package for the tarball?
https://golang.org/pkg/archive/tar/